### ISSUE TITLE
Be lenient with empty line in csv files

### ISSUE THREAD

**Issue by dmehra** [juttle/juttle#557]
Our csv file reader doesn't like files with empty lines (anywhere: top, middle, last line). Other csv readers out there accept such files just fine. We should be lenient about that too, as the problem isn't easy to identify for the user, given a large input file.

Sample file `/tmp/csv.csv`:
```

a,b,c,d

A,B,C,D

A,B,C,D

```

Juttle says (quite verbosely):
```
juttle> read file -file '/tmp/csv.csv' -format 'csv'
Error: internal error RuntimeError: Invalid CSV data: Error: Row length does not match headers { [RuntimeError: internal error RuntimeError: Invalid CSV data: Error: Row length does not match headers]
  message: 'internal error RuntimeError: Invalid CSV data: Error: Row length does not match headers',
  code: 'RT-INTERNAL-ERROR',
  info: 
   { error: 'RuntimeError: Invalid CSV data: Error: Row length does not match headers',
     procName: 'read',
     location: { filename: 'main', start: [Object], end: [Object] } } }
└───────────┴───────────┘
juttle> 
```

But for example [csvlint](https://github.com/Clever/csvlint) says
```
./csvlint /tmp/csv.csv 
file is valid
```

As far as the [RFC](https://tools.ietf.org/html/rfc4180) goes, it doesn't say anything one way or another about empty lines, only that the "last record in the file may or may not have an ending line break" (we handle that already). But it's a very bare RFC.

**Comment by dmajda** [comment]
Is this motivated by an actual problem (software or humans producing CSV file(s) with empty lines)?

> Other csv readers out there accept such files just fine.

Examples?

> As far as the [RFC](https://tools.ietf.org/html/rfc4180) goes, it doesn't say anything one way or another about empty lines, only that the "last record in the file may or may not have an ending line break" (we handle that already).

That RFC contains a grammar which pretty unambiguously specifies that there are no blank lines allowed — individual records can be separated only by one CRLF sequence:

```
file = [header CRLF] record *(CRLF record) [CRLF]
```

But the RFC is only informational (not on the standards track), so no big deal ignoring it.

**Comment by dmajda** [comment]
I filed #562 for the ugly error message.

**Comment by dmehra** [comment]
> Is this motivated by an actual problem (software or humans producing CSV file(s) with empty lines)?

Yes, I ran into this when processing a real CSV file from our internal confluence. I don't know what software or human had generated it.

> Other csv readers out there accept such files just fine.
> Examples?

The Numbers app on the Mac, csvlint tool I referenced above - I  tried using it, and some online linter, to find out what was wrong with my input file, but they treated it as valid, so I couldn't identify the problem without looking through the (pretty big) file by hand.

**Comment by dmehra** [comment]
For completeness, I've also encountered many CSV files on the web that end in an empty line followed by a line of invalid CSV that's meant to be a comment. For example, this wonderful source of statistical data is all line that (try any dataset, go to export - csv). https://datamarket.com/data/list/?q=provider:tsdl
I haven't asked to tolerate that format as it's clearly invalid CSV, and tools like csvlint properly identify the issue, for example:

```
csvlint ~/Downloads/monthly-milk-production-pounds-p.csv 
Record #169 has error: wrong number of fields in line
```

The file in question looks like this (skipping the middle):

```
"Month","Monthly milk production: pounds per cow. Jan 62 ? Dec 75"
"1962-01",589
"1962-02",561
"1975-12",843

Monthly milk production: pounds per cow. Jan 62 ? Dec 75

```

I would expect the user to fix up the file manually before feeding into Juttle in this case, as I did. However notice that here too, the linter only complains about the comment line, but not about the empty lines, so it's hard for the user to know that those should also be fixed up.

It doesn't help that Juttle error doesn't indicate which line number in the CSV file has invalid format, it would be better if we did that, like csvlint does.

**Comment by dmajda** [comment]
OK, all that makes sense, let’s follow Postel’s law (as the RFC recommends anyway) and ignore the empty (and probably also whitespace-only) lines.
